### PR TITLE
PP-120: variable amount of revenue partners

### PR DESCRIPTION
**Pull request by mortelli**
## What

This PR:
- improves upon the `Collector` contract and `ICollector` interface so that they can have a variable amount of revenue sharing partners
- removes the names for the revenue sharing partners (from `IOVLabsRecipient`, `WalletProvider` etc.)
  - in the addresses JSON file the partners will be shown as  `partner1`, `partner2`, etc.
- renames the owner parameter so that it does not imply a multisig address (using one or not should both be valid choices)
- as a piggyback change: adds the restriction of having `0` balance before being able to update shares, in order to avoid a wrongful distribution of tokens after an update
- as a piggyback change: adds the `transferOwnership` function

## Why

These changes are wanted to bring more flexibility to the collector smart contract and Revenue Sharing feature.

## Costs

Local tests were ran to compare costs between the two implementations. Below is a costs table, in gas units:

| Action | Original implementation (4 fixed partners) | New implementation (2 partners) | New implementation (4 partners) | New implementation (8 partners) |
| -------- | -------- | -------- | -------- | -------- |
| Collector contract deployment | 1.042.579 | 859.841 | 955.451 | 1.146.671 |
| Withdraw | 133.649, 73.649 | 73.420, 43.420 | 137.410, 77.410 | 265.391, 145.391 |
| Shares update | 70.865 | 182.238 | 142.802 | 159.465 |
| Ownership transfer | 28.533 | 28.511 | 28.511 | 28.511 |

Notes:
- All withdraws were executed with 10 tokens in the Collector contract. Some quick tests with 100 tokens yielded the same gas costs.
  - All `withdraw` receivers have RBTC but no token balance.
- There are two values for transfer costs since the first transfer is more expensive than the rest. Subsequent transfers all cost the same once executed (see **Conclusions & thoughts** section down below)
- All shares updates used the same parameter: 4 partners (the same fixed ones present in the current `revenue-sharing-POC` branch)
- No multisig owners were used for any of these tests.
- All tests were executed exclusively through truffle; the SDK dapp was not used (see **Notes** section down below)
- The ownership transfer code was copied locally to the previous implementation in order to run the tests and compare the costs for this action

### Conclusions & thoughts
- Strange behavior continues to occur related to `withdraw` costs. This might be linked to [Jira PP-117](https://rsklabs.atlassian.net/browse/PP-117).
- The deployment for the original contract seems surprisingly expensive. It might have to do with the types of variables used (old implementation uses a map and has names, for example) although this is not entirely clear.
- Deployment costs for the new Collector contract will increase the more partners are defined.
- `withdraw` costs for the new Collector contract will increase the more partners are defined, although they are roughly equivalent to the old implementation when using 4.
- The new share update is at least twice as expensive as the previous implementation.
- Shares update costs for the new implementation depend on both the original size of the partners as well as the size of the new set that will replace them.
- Ownership transfer cost is equivalent across all cases, which is expected.

## Notes

This PR will break both the `withdraw` script as well as possibly the SDK dapp in their current form (`revenue-sharing-POC` branches). These can either be updated before merging this PR, or updated right after.

## Refs

- [Jira PP-120](https://rsklabs.atlassian.net/browse/PP-120)

### CODE REVIEW

File: contracts/Collector.sol [hunk 1]
```diff
@@ -7,67 +7,70 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+    function updateShares(RevenuePartner[] memory _partners) 
     public
-    validShares(_shares)
-    onlyMultisigOwner()
+    validShares(_partners)
+    onlyOwner()
     {
-        revenueShares = _shares;
+        uint balance = token.balanceOf(address(this));
+        require(balance == 0, "can't update with balance > 0");
```
[the remark this answers is not in the document]
Comment (antomor) [reply contracts/Collector.sol:34]: Yes, totally agree. Thank you for this change

File: contracts/Collector.sol [hunk 2]
```diff
@@ -7,67 +7,70 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+    
+        delete partners;
+        
+        for (uint i = 0; i < _partners.length; i++)
+            partners.push(_partners[i]);
```
[the remark this answers is not in the document]
Comment (antomor) [reply contracts/Collector.sol:39]: Not sure, what do you mean by this? We are executing the `validShares` modifier, aren't we?

File: migrations/3_revenue_sharing.js [hunk 3]
```diff
@@ -66,19 +55,15 @@ module.exports = async function (deployer, network) {
...
-        liquidityProviderShare: liquidityProviderShare,
-        iovLabsRecipientShare: iovLabsRecipientShare,
         collectorContract: Collector.address,
-        collectorSafe: await collectorInstance.multisigOwner.call(),
+        collectorOwner: await collectorInstance.owner.call(),
         collectorToken: await collectorInstance.token.call(),
     };
 
+    revenueSharingPartners.forEach(function (partner, i) {
+        jsonConfig[networkId]["partner" + (i+1)] = partner.beneficiary;
```
Comment (antomor) [inline migrations/3_revenue_sharing.js:64]: if possible, I would generate a JSON with a different structure for the partner parts:
```
{
   ...
   "partner1": {
        "address": "0x..."
        "share": "..."
   }
   ...
}
```

**Review by antomor — COMMENTED** [review]
Thank you for this PR. It looks good to me, I've just left some comments here and there.

Regarding this:
> This PR will break both the withdraw script as well as possibly the SDK dapp in their current form (revenue-sharing-POC branches). These can either be updated before merging this PR, or updated right after.

I'd prefer to include the changes of the withdraw script in this PR. If you prefer I can take care of it.

**Side note**:
WDYT if we decide to support two different collector implementations? I mean, we could develop a collector implementation that requires that the owner is a multisig address and a simpler version (the one you're proposing right now). Maybe the multisig version could be tackled on a different PR/task.

File: migrations/3_revenue_sharing.js (hunk 3 above)
Comment (mortelli) [reply migrations/3_revenue_sharing.js]: much better! thanks, fixed

**Pull request merged**